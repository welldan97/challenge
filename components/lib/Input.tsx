import { useField } from 'formik';
import React, { memo } from 'react';
import styled from 'styled-components';

// SECTION: Styles

const Wrapper = styled.div`
  padding-left: 24px;
  padding-right: 24px;
  height: 88px;
`;

const Label = styled.label`
  display: block;
  margin-bottom: 6px;
  margin-left: 8px;
`;

const BaseInput = styled.input`
  height: 44px;
  background: #ffffff;
  border: 1px solid #6932d4;
  border-radius: 8px;
  padding: 8px;
  width: 100%;
`;

// !SECTION
// SECTION: Main

interface Props {
  name: string;
  placeholder: string;
  label: string;
}

const Input = memo(({ name, label, placeholder }: Props) => {
  const [field] = useField({
    name,
  });

  return (
    <Wrapper>
      <Label htmlFor={name}>{label}</Label>
      <BaseInput
        value={field.value}
        id={name}
        placeholder={placeholder}
        onChange={field.onChange}
        onBlur={field.onBlur}
      />
    </Wrapper>
  );
});

export default Input;
