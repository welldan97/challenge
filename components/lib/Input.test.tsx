import { Formik } from 'formik';
import { render, screen, fireEvent, act } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import Input from './Input';

// SECTION: Main

describe('Input', () => {
  test('render', () => {
    const initialValues = {
      testName: 'testValue',
    };

    const onSubmit = () => {};

    const { container } = render(
      <Formik initialValues={initialValues} onSubmit={onSubmit}>
        {() => (
          <Input
            name="testName"
            label="TEST_LABEL"
            placeholder="TEST_PLACEHOLDER"
          />
        )}
      </Formik>,
    );

    expect(container).toMatchSnapshot();
  });

  test('change value', async () => {
    const initialValues = {
      testName: 'TEST_VALUE',
    };

    const onSubmit = jest.fn();

    render(
      <Formik initialValues={initialValues} onSubmit={onSubmit}>
        {({ handleSubmit }) => (
          <form onSubmit={handleSubmit} data-testid="form">
            <Input
              name="testName"
              label="TEST_LABEL"
              placeholder="TEST_PLACEHOLDER"
            />
          </form>
        )}
      </Formik>,
    );

    const input = screen.getByTestId('testName');
    await userEvent.clear(input);
    await userEvent.type(input, 'NEW_VALUE');

    await act(() => {
      fireEvent.blur(input);
      const form = screen.getByTestId('form');

      fireEvent.submit(form);
    });

    expect(onSubmit.mock.calls[0][0]).toEqual({
      testName: 'NEW_VALUE',
    });
  });
});
