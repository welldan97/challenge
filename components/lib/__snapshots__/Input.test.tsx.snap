// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input render 1`] = `
<div>
  <div
    class="sc-2d2a1aa8-0 cxcNIV"
  >
    <label
      class="sc-2d2a1aa8-1 haCvit"
      for="testName"
    >
      TEST_LABEL
    </label>
    <input
      class="sc-2d2a1aa8-2 eQinJl"
      data-testid="testName"
      id="testName"
      placeholder="TEST_PLACEHOLDER"
      value="testValue"
    />
  </div>
</div>
`;
